File: src/adapters/trivy.ts
import type { NormalizedVulnerability, NormalizedPackage } from '../types/index.js';
import { mapSeverity } from './severity.js';

function formatLicense(license: any): string | undefined {
  if (!license) return undefined;
  if (typeof license === 'string') return license;
  if (Array.isArray(license)) {
    const formatted = license.map((l) => formatLicense(l)).filter(Boolean);
    return formatted.length > 0 ? formatted.join(', ') : undefined;
  }
  if (typeof license === 'object') {
    if (license.value) return license.value;
    if (license.spdxExpression) return license.spdxExpression;
    if (license.name) return license.name;
    if (license.license) return license.license;
    if (license.expression) return license.expression;
    const values = Object.values(license);
    const firstString = values.find((v) => typeof v === 'string' && v !== 'declared');
    if (firstString) return firstString as string;
  }
  return undefined;
}

export function extractTrivyVulnerabilities(report: any): NormalizedVulnerability[] {
  const findings: NormalizedVulnerability[] = [];

  if (report?.Results) {
    for (const result of report.Results) {
      if (result.Vulnerabilities) {
        for (const vuln of result.Vulnerabilities) {
          findings.push({
            source: 'trivy',
            cveId: vuln.VulnerabilityID || vuln.PkgID,
            packageName: vuln.PkgName || vuln.PkgID,
            installedVersion: vuln.InstalledVersion ?? undefined,
            fixedVersion: vuln.FixedVersion ?? undefined,
            severity: mapSeverity(vuln.Severity) as NormalizedVulnerability['severity'],
            cvssScore: vuln.CVSS?.nvd?.V3Score || vuln.CVSS?.redhat?.V3Score || undefined,
            vulnerabilityUrl: vuln.PrimaryURL ?? undefined,
            title: vuln.Title ?? undefined,
            description: vuln.Description ?? undefined,
          });
        }
      }
    }
  }

  return findings;
}

export function extractTrivyPackages(report: any): NormalizedPackage[] {
  const findings: NormalizedPackage[] = [];

  if (report?.Results) {
    for (const result of report.Results) {
      if (result.Packages) {
        for (const pkg of result.Packages) {
          findings.push({
            source: 'trivy',
            name: pkg.Name,
            version: pkg.Version || '',
            type: result.Type || 'unknown',
            license: formatLicense(pkg.License),
          });
        }
      }
    }
  }

  return findings;
}

File: internal/adapter/trivy.go
package adapter

import (
	"github.com/HarborGuard/harborguard-sensor/internal/types"
)

// ExtractTrivyVulnerabilities extracts normalized vulnerabilities from trivy JSON.
func ExtractTrivyVulnerabilities(report interface{}) []types.NormalizedVulnerability {
	var findings []types.NormalizedVulnerability

	data, ok := report.(map[string]interface{})
	if !ok {
		return findings
	}

	results, ok := data["Results"].([]interface{})
	if !ok {
		return findings
	}

	for _, r := range results {
		result, ok := r.(map[string]interface{})
		if !ok {
			continue
		}
		vulns, ok := result["Vulnerabilities"].([]interface{})
		if !ok {
			continue
		}
		for _, v := range vulns {
			vuln, ok := v.(map[string]interface{})
			if !ok {
				continue
			}

			cveID := getString(vuln, "VulnerabilityID")
			if cveID == "" {
				cveID = getString(vuln, "PkgID")
			}
			pkgName := getString(vuln, "PkgName")
			if pkgName == "" {
				pkgName = getString(vuln, "PkgID")
			}

			var cvssScore *float64
			if cvss, ok := vuln["CVSS"].(map[string]interface{}); ok {
				if nvd, ok := cvss["nvd"].(map[string]interface{}); ok {
					if s, ok := nvd["V3Score"].(float64); ok {
						cvssScore = &s
					}
				}
				if cvssScore == nil {
					if rh, ok := cvss["redhat"].(map[string]interface{}); ok {
						if s, ok := rh["V3Score"].(float64); ok {
							cvssScore = &s
						}
					}
				}
			}

			findings = append(findings, types.NormalizedVulnerability{
				Source:           "trivy",
				CveID:            cveID,
				PackageName:      pkgName,
				InstalledVersion: getString(vuln, "InstalledVersion"),
				FixedVersion:     getString(vuln, "FixedVersion"),
				Severity:         MapSeverity(getString(vuln, "Severity")),
				CvssScore:        cvssScore,
				VulnerabilityURL: getString(vuln, "PrimaryURL"),
				Title:            getString(vuln, "Title"),
				Description:      getString(vuln, "Description"),
			})
		}
	}

	return findings
}

// ExtractTrivyPackages extracts normalized packages from trivy JSON.
func ExtractTrivyPackages(report interface{}) []types.NormalizedPackage {
	var findings []types.NormalizedPackage

	data, ok := report.(map[string]interface{})
	if !ok {
		return findings
	}

	results, ok := data["Results"].([]interface{})
	if !ok {
		return findings
	}

	for _, r := range results {
		result, ok := r.(map[string]interface{})
		if !ok {
			continue
		}
		resultType := getString(result, "Type")
		if resultType == "" {
			resultType = "unknown"
		}

		pkgs, ok := result["Packages"].([]interface{})
		if !ok {
			continue
		}
		for _, p := range pkgs {
			pkg, ok := p.(map[string]interface{})
			if !ok {
				continue
			}
			version := getString(pkg, "Version")

			findings = append(findings, types.NormalizedPackage{
				Source:  "trivy",
				Name:    getString(pkg, "Name"),
				Version: version,
				Type:    resultType,
				License: formatLicense(pkg["License"]),
			})
		}
	}

	return findings
}

// formatLicense handles string, array, and object license formats.
func formatLicense(license interface{}) string {
	if license == nil {
		return ""
	}
	switch v := license.(type) {
	case string:
		return v
	case []interface{}:
		var parts []string
		for _, l := range v {
			if s := formatLicense(l); s != "" {
				parts = append(parts, s)
			}
		}
		if len(parts) > 0 {
			return joinStrings(parts, ", ")
		}
		return ""
	case map[string]interface{}:
		for _, key := range []string{"value", "spdxExpression", "name", "license", "expression"} {
			if s, ok := v[key].(string); ok && s != "" {
				return s
			}
		}
		// Find first non-"declared" string value
		for _, val := range v {
			if s, ok := val.(string); ok && s != "declared" {
				return s
			}
		}
	}
	return ""
}

func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func joinStrings(parts []string, sep string) string {
	result := ""
	for i, p := range parts {
		if i > 0 {
			result += sep
		}
		result += p
	}
	return result
}
